### ISSUE TITLE
NETCONF session persistence

### ISSUE THREAD

**Issue by mdoshi-com** [CESNET/netopeer2#722]
Hi,
NETCONF sessions should be long lived, i.e persisted unless either side closes or some error happens (refer https://tools.ietf.org/html/rfc6241#section-2.1)
How is the session kept alive even when the session is idle?
Does netopeer2-server support this feature?

Regards,
Meghna

**Comment by michalvasko** [comment]
Hi,
the persistence is guaranteed by TCP and its keep-alive feature is also used.

**Comment by mdoshi-com** [comment]
Hi Michal,
Is there any way by which application can control the keep alive periodicity for the session alone, without changing the tcp keep alive for the whole system?

Regards,
Meghna

**Comment by michalvasko** [comment]
Yes, there are specific [YANG nodes](https://github.com/CESNET/netopeer2/blob/master/modules/ietf-tcp-common%402019-07-02.yang#L64) for this. Print `ietf-netconf-server` module in the `tree` format to actually see where the values are in the configuration.

**Comment by mdoshi-com** [comment]
Hi Michal,
Ok yes got it. Thanks.

**Comment by iGovindY** [comment]
Hi Michal,

As you suggested, we were verifying these configuration behaviour on netopeer2

**configured below for tcp keepalives**

```
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters (container)
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/local-address = (STRING) 10.207.50.21
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/keepalives (container)
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/keepalives/idle-time = (UINT16) 60
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/keepalives/max-probes = (UINT16) 3
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/keepalives/probe-interval = (UINT16) 5
/ietf-netconf-server:netconf-server/listen/endpoint[name='default-ssh']/ssh/tcp-server-parameters/local-port = (UINT16) 830 [default]
```

The server is running on 10.207.50.21 and client is running on 10.207.50.26.
What we are seeing is netopeer2-server is not behaving as per these configurations please see the below image

![image](https://user-images.githubusercontent.com/29160257/95956174-8d274900-0e1b-11eb-8be3-5c3cab312b27.png)
![image](https://user-images.githubusercontent.com/29160257/95956861-8816c980-0e1c-11eb-97e9-17c33f312724.png)
![image](https://user-images.githubusercontent.com/29160257/95956974-b1cff080-0e1c-11eb-8256-92a8fe3d59d9.png)

I can provide tcpdump file if needed.

Also one more query:
1. will the netconf client also send keepalives based on above configurations? or client has no relation to this configurations.

Thanks
Govind.

**Comment by michalvasko** [comment]
So why do you think it is not working? I can only see that idle time is not really respected and for some reason it is 3.5s (when keep-alive packets starts to be sent after last sent TCP packet). Why that is I cannot tell you, all that is performed is [this](https://github.com/CESNET/libnetconf2/blob/master/src/session.c#L153) system call, all the rest is up to the kernel.

As for the client, no, these are the options set for server. They can be set for the client as well (*libnetconf2* supports it), but *netopeer2-cli* does not support it.
